src/Day12/Solution.hs

```diff
@@ -0,0 +1,131 @@
+    go state (R, 90) = state {sHeadingV1 = (next . sHeadingV1) state}
+    go state (R, 180) = state {sHeadingV1 = (next . next . sHeadingV1) state}
+    go state (R, 270) = state {sHeadingV1 = (next . next . next . sHeadingV1) state}
+    go state (L, 90) = state {sHeadingV1 = (prev . sHeadingV1) state}
+    go state (L, 180) = state {sHeadingV1 = (prev . prev . sHeadingV1) state}
+    go state (L, 270) = state {sHeadingV1 = (prev . prev . prev . sHeadingV1) state}
+    go state (F, value) = state {sPositionV1 = nextPoint (sHeadingV1 state) (sPositionV1 state) value}
+    go state (N, value) = state {sPositionV1 = nextPoint North (sPositionV1 state) value}
+    go state (E, value) = state {sPositionV1 = nextPoint East (sPositionV1 state) value}
+    go state (S, value) = state {sPositionV1 = nextPoint South (sPositionV1 state) value}
+    go state (W, value) = state {sPositionV1 = nextPoint West (sPositionV1 state) value}
+    go _ _ = error "Unknown action"
+
+    nextPoint :: Direction -> Point -> Value -> Point
+    nextPoint direction point value = case direction of
+      North -> moveY value point
+      East -> moveX value point
+      South -> moveY (- value) point
+      West -> moveX (- value) point
+
+    next :: (Eq p, Bounded p, Enum p) => p -> p
+    next e
+      | e == maxBound = minBound
+      | otherwise = succ e
+
+    prev :: (Eq p, Bounded p, Enum p) => p -> p
+    prev e
+      | e == minBound = maxBound
+      | otherwise = pred e
+
+data StateV2 = StateV2 {sWaypointV2 :: Point, sPositionV2 :: Point} deriving (Show)
+
+runV2 :: [Instruction] -> Point
+runV2 = sPositionV2 . foldl go initialState
+  where
+    initialState :: StateV2
+    initialState = StateV2 (Point 10 1) (Point 0 0)
+
+    go :: StateV2 -> Instruction -> StateV2
+    go state (R, 90) = state {sWaypointV2 = rotateRight . sWaypointV2 $ state}
+    go state (R, 180) = state {sWaypointV2 = rotateRight . rotateRight . sWaypointV2 $ state}
+    go state (R, 270) = state {sWaypointV2 = rotateRight . rotateRight . rotateRight . sWaypointV2 $ state}
+    go state (L, 90) = state {sWaypointV2 = rotateLeft . sWaypointV2 $ state}
+    go state (L, 180) = state {sWaypointV2 = rotateLeft . rotateLeft . sWaypointV2 $ state}
+    go state (L, 270) = state {sWaypointV2 = rotateLeft . rotateLeft . rotateLeft . sWaypointV2 $ state}
+    go state (F, value) =
+      state
+        { sPositionV2 =
+            state
+              & sPositionV2
+              & moveY ((* value) . getY . sWaypointV2 $ state)
+              & moveX ((* value) . getX . sWaypointV2 $ state)
+        }
+    go state (N, value) = state {sWaypointV2 = moveY value . sWaypointV2 $ state}
+    go state (E, value) = state {sWaypointV2 = moveX value . sWaypointV2 $ state}
+    go state (S, value) = state {sWaypointV2 = moveY (- value) . sWaypointV2 $ state}
+    go state (W, value) = state {sWaypointV2 = moveX (- value) . sWaypointV2 $ state}
+    go _ _ = error "Unknown action"
```

Comment:
For my goals, I'm interested in finding good abstractions between part 1 and part 2.  This solution feels like "tapping out" because I'm rewriting a totally different mechanism from the one used in part 1.


Having thought about it a little bit, I believe part 1 could be solved using waypoints, which would make it be more in line with part 2. 

TODO: 

- [ ] take a  second look at finding a good abstraction between part 1 and part 2